gobblin-data-management/src/main/java/gobblin/data/management/copy/CopySource.java

```diff
@@ -88,57 +101,128 @@
   @Override
   public List<WorkUnit> getWorkunits(SourceState state) {
 
-    List<WorkUnit> workUnits = Lists.newArrayList();
     CopyContext copyContext = new CopyContext();
 
     try {
 
+      FileSystem sourceFs = getSourceFileSystem(state);
+      FileSystem throttledSourceFs = getOptionallyThrottledFileSystem(sourceFs, state);
+      FileSystem targetFs = getTargetFileSystem(state);
+      FileSystem throttledTargetFs = getOptionallyThrottledFileSystem(targetFs, state);
+
       DatasetFinder<CopyableDataset> datasetFinder =
-          DatasetUtils.instantiateDatasetFinder(state.getProperties(), getSourceFileSystem(state),
-              DEFAULT_DATASET_PROFILE_CLASS_KEY);
+          DatasetUtils.instantiateDatasetFinder(state.getProperties(), sourceFs, DEFAULT_DATASET_PROFILE_CLASS_KEY);
       List<CopyableDataset> copyableDatasets = datasetFinder.findDatasets();
-      FileSystem targetFs = getTargetFileSystem(state);
 
-      for (CopyableDataset copyableDataset : copyableDatasets) {
+      // TODO: The comparator sets the priority of file sets. Currently, all file sets have the same priority, this needs to
+      // be pluggable.
+      ConcurrentBoundedWorkUnitList workUnitList = new ConcurrentBoundedWorkUnitList(100000,
```

Comment:
Should we keep 100000 as the default size and make it configurable? If not we can have a no args constructors and have the CouncurrentBoundedWorkUnitList assume size and comparator